chat_v2_message_counter_sample/src/main/java/com/zendesk/chat_v2/sample/UnreadMessageCounter.java

```diff
@@ -0,0 +1,73 @@
+package com.zendesk.chat_v2.sample;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import zendesk.chat.Chat;
+import zendesk.chat.ChatLog;
+import zendesk.chat.ChatState;
+import zendesk.chat.ObservationScope;
+import zendesk.chat.Observer;
+
+class UnreadMessageCounter {
+
+    private final UnreadMessageCounterListener unreadMessageCounterListener;
+
+    private boolean shouldCount;
+    private String lastChatLogId;
+    private String lastReadChatLogId;
+    private List<ChatLog> unreadChatLogs = new ArrayList<>();
+    private ObservationScope observationScope = new ObservationScope();
+
+    public UnreadMessageCounter(UnreadMessageCounterListener unreadMessageCounterListener) {
+        this.unreadMessageCounterListener = unreadMessageCounterListener;
+
+        Chat.INSTANCE.providers().chatProvider().observeChatState(observationScope, new Observer<ChatState>() {
+            @Override
+            public void update(ChatState chatState) {
+                if (chatState != null && !chatState.getChatLogs().isEmpty()) {
+                    if (shouldCount && lastReadChatLogId !=  null) {
+                        updateCounter(chatState.getChatLogs());
+                    }
+                    lastChatLogId = chatState.getChatLogs().get(
+                            chatState.getChatLogs().size() - 1
+                    ).getId();
+                }
+            }
+        });
+    }
+
+
+    // Determines whether or not the chat websocket should be re-connected to.
+    // A connected, open websocket will disable push notifications and receive messages as normal.
+    public void startCounter() {
+        shouldCount = true;
+        lastReadChatLogId = lastChatLogId;
+        Chat.INSTANCE.providers().connectionProvider().connect();
+    }
+
+    public void stopCounter() {
+        unreadChatLogs = new ArrayList<>();
+        lastReadChatLogId = null;
+        shouldCount = false;
+        unreadMessageCounterListener.onUnreadCountUpdated(0);
+    }
+
+    // Increment the counter and send an update to the listener
+    synchronized private void updateCounter(List<ChatLog> chatLogs) {
```

Comment:
For better synchonization (which may or may not be a real term) it might be better to pass in the `lastReadIndex` here rather than referencing the field, as theoretically it could be changed from a different thread.